Optional enum match C error

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.4 754c83a
OS: linux, "Arch Linux"
Processor: 16 cpus, 64bit, little endian, AMD Ryzen 7 PRO 4750U with Radeon Graphics

getwd: /home/hector
vexe: /home/hector/Dev/v/v
vexe mtime: 2024-01-26 11:31:42

vroot: OK, value: /home/hector/Dev/v
VMODULES: OK, value: /home/hector/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.43.0
Git vroot status: weekly.2024.04-11-g754c83a8-dirty
.git/config present: true

CC version: cc (GCC) 13.2.1 20230801
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3
```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg test.v`
```v
enum Test {
    a
    b
}

fn optional() ?Test {
    return none
}

fn main() {
    match optional() {
        .a { }
        else { }
    }
}
```

**What did you expect to see?**

No output

**What did you see instead?**
```
==================
/tmp/v_1000/test.01HN3750AMESTN4YW9B51T59G9.tmp.c:12862: error: invalid operand types for binary operation
/tmp/v_1000/test.01HN3750AMESTN4YW9B51T59G9.tmp.c:12862: error: invalid aggregate type for register load
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/match.v
+++ b/vlib/v/checker/match.v
@@ -55,6 +55,10 @@ fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 					branch.branch_pos)
 			}
 		}
+		if !branch.is_else && cond_is_option && branch.exprs[0] !is ast.None {
+			c.error('`match` expression with Option type only checks against `none`, to match its value you must unwrap it first `var?`',
+				branch.pos)
+		}
 		// If the last statement is an expression, return its type
 		if branch.stmts.len > 0 {
 			mut stmt := branch.stmts.last()
@@ -63,10 +67,6 @@ fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 					c.expected_type = node.expected_type
 				}
 				expr_type := c.expr(mut stmt.expr)
-				if !branch.is_else && cond_is_option && branch.exprs[0] !is ast.None {
-					c.error('`match` expression with Option type only checks against `none`, to match its value you must unwrap it first `var?`',
-						branch.pos)
-				}
 				stmt.typ = expr_type
 				if first_iteration {
 					if node.is_expr && (node.expected_type.has_flag(.option)

--- a/vlib/v/checker/tests/option_fn_err.out
+++ b/vlib/v/checker/tests/option_fn_err.out
@@ -3,21 +3,21 @@ vlib/v/checker/tests/option_fn_err.vv:34:16: error: cannot use `?int` as `int`, 
    33 |     _ := bar(0)
    34 |     println(twice(bar(0)))
       |                   ~~~~~~
-   35 |
+   35 | 
    36 |     // anon fn
 vlib/v/checker/tests/option_fn_err.vv:37:16: error: cannot use `?int` as `int`, it must be unwrapped first in argument 1 to `anon`
-   35 |
+   35 | 
    36 |     // anon fn
    37 |     fn (_ int) {}(bar(0))
       |                   ~~~~~~
-   38 |
+   38 | 
    39 |     // assert
 vlib/v/checker/tests/option_fn_err.vv:40:9: error: assert can be used only with `bool` expressions, but found `bool` instead
-   38 |
+   38 | 
    39 |     // assert
    40 |     assert bar(true)
       |            ~~~~~~~~~
-   41 |
+   41 | 
    42 |     // struct
 vlib/v/checker/tests/option_fn_err.vv:45:3: error: cannot assign an Option value to a non-option struct field
    43 |     mut v := Data{
@@ -38,7 +38,7 @@ vlib/v/checker/tests/option_fn_err.vv:49:6: error: cannot use `?int` as `int`, i
    48 |     v.add(bar(0)) // call method
    49 |     v.f(bar(0)) // call fn field
       |         ~~~~~~
-   50 |
+   50 | 
    51 |     // array
 vlib/v/checker/tests/option_fn_err.vv:53:6: error: unwrapped Option cannot be used in an infix expression
    51 |     // array
@@ -115,5 +115,12 @@ vlib/v/checker/tests/option_fn_err.vv:69:18: error: type mismatch, `bar` must re
    68 |     println(arr.any(bar(true)))
    69 |     println(arr.all(bar(true)))
       |                     ~~~~~~~~~
-   70 |
+   70 | 
    71 |     match bar(0) {
+vlib/v/checker/tests/option_fn_err.vv:72:3: error: `match` expression with Option type only checks against `none`, to match its value you must unwrap it first `var?`
+   70 | 
+   71 |     match bar(0) {
+   72 |         0 {}
+      |         ^
+   73 |         else {}
+   74 |     }

